libraries/AP_Camera/AP_RunCam.cpp

```diff
@@ -0,0 +1,831 @@
+    case RCDEVICE_CAM_KEY_LEFT:
+    case RCDEVICE_CAM_KEY_UP:
+    case RCDEVICE_CAM_KEY_RIGHT:
+    case RCDEVICE_CAM_KEY_DOWN:
+        simulate_5_key_OSD_cable_button_press(map_key_to_protocol_operation(key), FUNCTOR_BIND_MEMBER(&AP_RunCam::handle_5_key_simulation_response, void, RunCamResponse*));
+        break;
+    case RCDEVICE_CAM_KEY_RELEASE:
+        simulate_5_key_OSD_cable_button_release(FUNCTOR_BIND_MEMBER(&AP_RunCam::handle_5_key_simulation_response, void, RunCamResponse*));
+        break;
+    case RCDEVICE_CAM_KEY_NONE:
+    default:
+        break;
+    }
+}
+
+// every time we run the OSD menu simulation it's necessary to open the connection
+void AP_RunCam::open_5_key_OSD_cable_connection(rcdevice_resp_parse_func_t parseFunc)
+{
+    send_request_and_waiting_response(RCDEVICE_PROTOCOL_COMMAND_5KEY_CONNECTION,
+        RCDEVICE_PROTOCOL_5KEY_CONNECTION_OPEN, 400, 2, NULL, parseFunc);
+}
+
+// every time we exit the OSD menu simulation it's necessary to close the connection
+void AP_RunCam::close_5_key_OSD_cable_connection(rcdevice_resp_parse_func_t parseFunc)
+{
+    send_request_and_waiting_response(RCDEVICE_PROTOCOL_COMMAND_5KEY_CONNECTION,
+        RCDEVICE_PROTOCOL_5KEY_CONNECTION_CLOSE, 400, 2, NULL, parseFunc);
+}
+
+// simulate button press event of 5 key OSD cable with special button
+void AP_RunCam::simulate_5_key_OSD_cable_button_press(uint8_t operation, rcdevice_resp_parse_func_t parseFunc)
+{
+    if (operation == RCDEVICE_PROTOCOL_5KEY_SIMULATION_NONE) {
+        return;
+    }
+
+    send_request_and_waiting_response(RCDEVICE_PROTOCOL_COMMAND_5KEY_SIMULATION_PRESS, operation, 400, 2, NULL, parseFunc);
+}
+
+// simulate button release event of 5 key OSD cable
+void AP_RunCam::simulate_5_key_OSD_cable_button_release(rcdevice_resp_parse_func_t parseFunc)
+{
+    send_request_and_waiting_response(RCDEVICE_PROTOCOL_COMMAND_5KEY_SIMULATION_RELEASE,
+        RCDEVICE_PROTOCOL_5KEY_SIMULATION_NONE, 400, 2, NULL, parseFunc);
+}
+
+// send a RunCam request and register a response to be processed
+void AP_RunCam::send_request_and_waiting_response(uint8_t commandID, uint8_t param,
+    uint32_t timeout, uint16_t maxRetryTimes, void *userInfo, rcdevice_resp_parse_func_t parserFunc)
+{
+    flush_rx_buffer();
+
+    RunCamResponse* response = new RunCamResponse(this, commandID, param, timeout, maxRetryTimes, userInfo, parserFunc);
+
+    if (_waiting_response_queue.push(response)) {
+        debug("sending command: %d, op: %d\n", commandID, param);
+        // send packet
+        send_packet(commandID, param);
+    }
+}
+
+// send a packet to the serial port
+void AP_RunCam::send_packet(uint8_t command, uint8_t param)
+{
+    // is this device open?
+    if (!uart) {
+        return;
+    }
+
+    uint8_t *b = _buffer;
+    uint8_t param_len = param > 0 || command == RCDEVICE_PROTOCOL_COMMAND_CAMERA_CONTROL;
+
+    *b++ = RCDEVICE_PROTOCOL_HEADER;
+    *b++ = command;
+    if (param_len > 0) {
```

Comment:
OK, this line confused things, really.  I think things might be clearer to have a `have_param` boolean...